tests/ballerina-test/src/test/java/org/ballerinalang/test/services/dispatching/UriMatrixParametersMatchTest.java

```diff
@@ -0,0 +1,105 @@
+import org.wso2.transport.http.netty.message.HTTPCarbonMessage;
+import org.wso2.transport.http.netty.message.HttpMessageDataStreamer;
+
+public class UriMatrixParametersMatchTest {
+
+    private CompileResult application;
+
+    @BeforeClass()
+    public void setup() {
+        application = BServiceUtil
+                .setupProgramFile(this, "test-src/services/dispatching/uri-matrix-param-matching.bal");
+    }
+
+    @Test
+    public void testMatrixParamsAndQueryParamsMatching() {
+        String path = "/hello/t1/john;age=10;color=white/1991;month=may;day=12/foo;a=5;b=10?x=10&y=5";
+        HTTPTestRequest cMsg = MessageUtils.generateHTTPMessage(path, "GET");
+        HTTPCarbonMessage response = Services.invokeNew(application, cMsg);
+
+        Assert.assertNotNull(response, "Response message not found");
+        BJSON bJson = new BJSON(new HttpMessageDataStreamer(response).getInputStream());
+
+        Assert.assertEquals(bJson.value().get("pathParams").asText(), "john, 1991");
+        Assert.assertEquals(bJson.value().get("personMatrix").asText(), "age=10;color=white");
+        Assert.assertEquals(bJson.value().get("yearMatrix").asText(), "month=may;day=12");
+        Assert.assertEquals(bJson.value().get("fooMatrix").asText(), "a=5;b=10");
+        Assert.assertEquals(bJson.value().get("queryParams").asText(), "x=10&y=5");
+    }
+
+    @Test
+    public void testEncodedPathDispatching() {
+        String path = "/hello/t2/john;age=2;color=white/foo%3Ba%3D5%3Bb%3D10"; // encoded URI
+        HTTPTestRequest cMsg = MessageUtils.generateHTTPMessage(path, "GET");
+        HTTPCarbonMessage response = Services.invokeNew(application, cMsg);
+
+        Assert.assertNotNull(response, "Response message not found");
+        BJSON bJson = new BJSON(new HttpMessageDataStreamer(response).getInputStream());
+
+        Assert.assertEquals(bJson.value().get("person").asText(), "john");
+        Assert.assertEquals(bJson.value().get("personParamSize").asText(), "2");
+        Assert.assertEquals(bJson.value().get("fooParamSize").asText(), "0");
+    }
+
+    @Test
+    public void testEncodedPathParamDispatching() {
+        String path = "/hello/t2/john%3Bage%3D2%3Bcolor%3Dwhite/foo%3Ba%3D5%3Bb%3D10"; // encoded URI
+        HTTPTestRequest cMsg = MessageUtils.generateHTTPMessage(path, "GET");
+        HTTPCarbonMessage response = Services.invokeNew(application, cMsg);
+
+        Assert.assertNotNull(response, "Response message not found");
+        BJSON bJson = new BJSON(new HttpMessageDataStreamer(response).getInputStream());
+
+        Assert.assertEquals(bJson.value().get("person").asText(), "john;age=2;color=white");
+        Assert.assertEquals(bJson.value().get("personParamSize").asText(), "0");
+        Assert.assertEquals(bJson.value().get("fooParamSize").asText(), "0");
+    }
+
+    @Test
+    public void testNonEncodedUrlDispatching() {
+        String path = "/hello/t2/john;age=2;color=white/foo;a=5;b=10"; // encoded URI
```

Comment:
Shall we test a path which doesn't have a consecutive matrix path segments?
i.e /hello/t2/john;age=2;color=white/product/foo;a=5;b=10"

Reply:
Added 
"/hello/t1/john;age=10;color=white/bar/1991;month=may;day=12/foo;a=5;b=10?x=10&y=5"